tikv/raftstore/snap.go

```diff
@@ -0,0 +1,834 @@
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	s.holdTmpFiles = false
+	return nil
+}
+
+func (s *Snap) Build(dbSnap *DBSnapshot, region *metapb.Region, snapData *rspb.RaftSnapshotData,
+	stat *SnapStatistics, deleter SnapshotDeleter) error {
+	if s.Exists() {
+		err := s.validate(dbSnap.db)
+		if err == nil {
+			return nil
+		}
+		log.Errorf("[region %d] file %s is corrupted, will rebuild: %v", region.Id, s.Path(), err)
+		if !retryDeleteSnapshot(deleter, s.key, s) {
+			log.Errorf("[region %d] failed to delete corrupted snapshot %s because it's already registered elsewhere",
+				region.Id, s.Path())
+			return err
+		}
+		err = s.initForBuilding()
+		if err != nil {
+			return err
+		}
+	}
+
+	builder, err := newSnapBuilder(s.CFFiles, dbSnap.db, dbSnap.lockStore, region)
+	if err != nil {
+		return err
+	}
+	err = builder.build()
+	if err != nil {
+		return err
+	}
+	log.Infof("region %d scan snapshot %s, key count %d, size %d", region.Id, s.Path(), builder.kvCount, builder.size)
+	err = s.saveCFFiles()
+	if err != nil {
+		return err
+	}
+	stat.KVCount = builder.kvCount
+	snapshotMeta, err := genSnapshotMeta(s.CFFiles)
+	if err != nil {
+		return err
+	}
+	s.MetaFile.Meta = snapshotMeta
+	err = s.saveMetaFile()
+	if err != nil {
+		return err
+	}
+	totalSize := s.TotalSize()
+	stat.Size = totalSize
+	// set snapshot meta data
+	snapData.FileSize = totalSize
+	snapData.Version = snapshotVersion
+	snapData.Meta = s.MetaFile.Meta
+	return nil
+}
+
+func (s *Snap) Path() string {
+	return s.displayPath
+}
+
+func (s *Snap) Exists() bool {
+	for _, cfFile := range s.CFFiles {
+		if cfFile.Size > 0 && !util.FileExists(cfFile.Path) {
+			return false
+		}
+	}
+	return util.FileExists(s.MetaFile.Path)
+}
+
+func (s *Snap) Delete() {
+	log.Debugf("deleting %s", s.Path())
+	for _, cfFile := range s.CFFiles {
+		_, err := util.DeleteFileIfExists(cfFile.ClonePath)
+		if err != nil {
+			panic(err)
+		}
+		if s.holdTmpFiles {
+			_, err = util.DeleteFileIfExists(cfFile.TmpPath)
+			if err != nil {
+				panic(err)
+			}
+		}
+		deleted, err := util.DeleteFileIfExists(cfFile.Path)
+		if err != nil {
+			panic(err)
+		}
+		if deleted {
+			atomic.AddInt64(s.SizeTrack, -int64(cfFile.Size))
+		}
+	}
+	_, err := util.DeleteFileIfExists(s.MetaFile.Path)
+	if err != nil {
+		panic(err)
+	}
+	if s.holdTmpFiles {
+		_, err := util.DeleteFileIfExists(s.MetaFile.TmpPath)
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func (s *Snap) Meta() (os.FileInfo, error) {
+	fi, err := os.Stat(s.MetaFile.Path)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	return fi, nil
+}
+
+func (s *Snap) TotalSize() (total uint64) {
+	for _, cf := range s.CFFiles {
+		total += cf.Size
+	}
+	return
+}
+
+func (s *Snap) Save() error {
+	log.Debugf("saving to %s", s.MetaFile.Path)
+	for _, cfFile := range s.CFFiles {
+		if cfFile.Size == 0 {
+			// skip empty cf file.
+			continue
+		}
+		// Check each cf file has been fully written, and the checksum matches.
```

Comment:
Don't need to flush explicitly?

Reply:
No, the `*os.FIle` doesn't even have the `flush` method.